Generate random verses

gazayas opened the pull request:
Hello,
I saw issue #38 which talks about retrieving random verses, so I decided to implement that here.
I don't think we want to get a truly random verse (something like Leviticus 22:17),
so I made a JSON file which holds verse references, and each reference is under a topic.

### Additions
1. Make the endpoint `/?random=verse` which retrieves a random verse.
2. Enable the users to choose a topic of their own by passing the topic name as a value in place of "verse".
i.e. - `/?random=hope`
3. Create a JSON file which holds all of the references. Each reference has its own topic.

### Changes
1. Encapsulate the database query logic into a method.

Since using the endpoint `/?random=verse` would call the `/` path, I put all of the database query logic into a method of its own and called it under `/` and `/:ref`, respectively.

Let me know what you think!
I'd be glad to look over the code and make any changes accordingly.

seven1m commented:
This is really cool, and thanks for your work!

One principle of bible-api.com is that it's just the raw data -- no commentary. I'm not convinced we should add a curated list of verses as I'm in no way qualified to maintain it. Nor am I sure who to trust to maintain it.

So I'm inclined to not accept this PR as-is for that reason.

I am open to discussion about this, though, as I also wonder what the utility of a purely random verse would be. I am aware some people make games using bible-api so perhaps it could be useful for that!

The code itself looks ok, with exception of your inconsistent indentation-style (4 spaces in multiple places).

I'll leave this open in case you want to discuss. If you want to rework it to return an actual purely random verse, then I think that would be acceptable and would allow us to close #38.

gazayas commented:
Thank you for the reply!

You're right about the spacing inconsistencies, I should have checked that before pushing the changes.
Since the PR is still open, I'll make a commit so the indentations are 2 spaces each.

I see what you're saying about the curated list of verses; although I wouldn't consider it a commentary, deciding which verses fall under which topic is relatively subjective, and I could see maintenance getting finicky. If we were to implement it, it would probably be best to use some solid base of information to refer to, like a third-party concordance of some sorts, but I think at at that point it just wouldn't be raw data anymore.

I'm impartial as to whether its topical or just raw data, so I'll see if I can come up with something that returns a purely random verse!

gazayas commented:
I've written some logic to generate a purely random verse this time.

It was a little tricky because the number of books vary from translation to translation.
Because the database queries I wrote here are similar to one another, I considered writing a method like `find_length` which would find the number of books, chapters, and then verses for a translation, but instead of writing some strange and convoluted regular expression to repeat, I just wrote each database query as is. They're relatively long, but I think the flow/intention is clear.

What I did decide to refactor was the 404 response when a translation isn't found.
The same code was already written in the original logic when trying to get a verse, so I didn't want to double it up in the `get_random_verse` method.

I think that's about it for the logic, but if there's anything I'm missing, just let me know.

seven1m reviewed (changes requested):
Lookin good! Just had a few small changes...

gazayas commented on app.rb:
```diff
@@ -30,24 +30,36 @@ def get_verses(ranges, translation_id)
-  end
+    translation = get_translation
+    return jsonp(translation[:error]) if translation[:error]
+    translation_id = DB['select id from translations where identifier = ?;', translation[:identifier]].first[:id]
```
Good call, I went ahead and just used `translation[:id]`.

gazayas commented:
Hello,
I've deleted the JSON file and adjusted the `translation_id` code.
I'm not sure if comments on the review send a notification, so I thought I'd just write a comment here too 👍

seven1m commented:
Sorry I've been busy, but I plan to take another peek at this in the next few days, merge it, and deploy! Thanks!

gazayas commented:
That was a blast 🚀
Thank you!

seven1m commented:
I just realized today that I never got around to deploying this. Yikes!

Sorry it took so long, but you can now use this on the site:

https://bible-api.com/?random=verse

gazayas commented:
Awesome!
